### ISSUE TITLE
resolved: does not remove answer section when dnssec validation failed

### ISSUE THREAD

**Issue by mateusz834** [systemd/systemd#22539]
**systemd version the issue has been seen with**
 > systemd 250 (250.3-4-arch)

**Used distribution**
 > arch linux

**Linux kernel version used** (`uname -a`)
 > `Linux arch 5.16.9-arch1-1 #1 SMP PREEMPT Fri, 11 Feb 2022 22:42:06 +0000 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
Empty answer section.
```
[mateusz@arch ~ ]$  dig @8.8.8.8 dnssec-failed.org 

; <<>> DiG 9.16.25 <<>> @8.8.8.8 dnssec-failed.org
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 29615
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;dnssec-failed.org.             IN      A

;; Query time: 256 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Wed Feb 16 20:16:46 CET 2022
;; MSG SIZE  rcvd: 46
```

**Unexpected behaviour you saw**
```
[mateusz@arch ~ ]$ resolvectl query dnssec-failed.org
dnssec-failed.org: resolve call failed: DNSSEC validation failed: missing-key
[mateusz@arch ~ ]$ dig @127.0.0.53 dnssec-failed.org 

; <<>> DiG 9.16.25 <<>> @127.0.0.53 dnssec-failed.org
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 26383
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;dnssec-failed.org.             IN      A

;; ANSWER SECTION:
dnssec-failed.org.      7077    IN      A       69.252.80.75

;; Query time: 10 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: Wed Feb 16 20:13:06 CET 2022
;; MSG SIZE  rcvd: 62
```

Resolved should only respond with bogus data when `cd` bit is set.
```
[mateusz@arch ~ ]$  dig @8.8.8.8 dnssec-failed.org +cd

; <<>> DiG 9.16.25 <<>> @8.8.8.8 dnssec-failed.org +cd
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 26566
;; flags: qr rd ra cd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;dnssec-failed.org.             IN      A

;; ANSWER SECTION:
dnssec-failed.org.      7200    IN      A       69.252.80.75

;; Query time: 146 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Wed Feb 16 20:18:34 CET 2022
;; MSG SIZE  rcvd: 62
```

**Comment by yuwata** [comment]
Hm, interesting. Once the domain is resolved with '+cd', then the subsequent queries include answer section.
```
$ resolvectl flush-caches
$ dig dnssec-failed.org

; <<>> DiG 9.16.28-RH <<>> dnssec-failed.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 35752
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;dnssec-failed.org.		IN	A

;; Query time: 1043 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: Sat May 07 15:39:33 JST 2022
;; MSG SIZE  rcvd: 46

$ dig +dnssec +cd dnssec-failed.org

; <<>> DiG 9.16.28-RH <<>> +dnssec +cd dnssec-failed.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 2396
;; flags: qr rd ra cd; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 65494
;; QUESTION SECTION:
;dnssec-failed.org.		IN	A

;; ANSWER SECTION:
dnssec-failed.org.	299	IN	A	96.99.227.255
dnssec-failed.org.	299	IN	RRSIG	A 5 2 300 20220521145106 20220504144606 44973 dnssec-failed.org. z+hx3iTpSjlUjt/AKgYPDyOPAYEmGTBUeuCDDpmn6sJoSRkmulrEXfPL dyzqBjVRYwrXM/8Jl/nWv+d1NwzszGq4S6rCYWVepb2yJxwYYWin1+9u EdTPa3/1q+2SA5amwBfmyRvbZAYnO73xtSIFEXesUmbAnPos0T5iDYy0 rgI=

;; Query time: 216 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: Sat May 07 15:40:05 JST 2022
;; MSG SIZE  rcvd: 239

$ dig +dnssec dnssec-failed.org

; <<>> DiG 9.16.28-RH <<>> +dnssec dnssec-failed.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 43019
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;dnssec-failed.org.		IN	A

;; ANSWER SECTION:
dnssec-failed.org.	253	IN	A	96.99.227.255

;; Query time: 13 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: Sat May 07 15:40:51 JST 2022
;; MSG SIZE  rcvd: 62


```

**Comment by rpigott** [comment]
Resolved by #31952 I think.

**Comment by poettering** [comment]
#31952 has been merged a while back. Closing.
